docs/design_notes/lift.md

```diff
@@ -0,0 +1,338 @@
+batch stats after apply and making sure state isn't mutable inside vmap when it's immutable outside (eval mode).
+
+
+We call the process of transforming a stateful Module into a pure function "functionalization".
+By temporarily turning a stateful Module into a function we make it compatible with JAX's functional transformations.
+
+## Lifting
+
+
+Flax provides an alternative for manual functionalization which we call lifted transformation.
+Lifted transformations are defined in `flax.core.lift`.
+All the lifted JAX transformations are defined with a single generic lifting API called `pack`.
+
+A number of decision decisions had to be made in order to define `pack`. The implementation
+of `pack` controls how variables and rngs are lifted and how fine-grained the user control is.
+It must also decide whether lifting decisions are made during variable definition or at the transformation.
+
+
+### Lifting granularity
+
+
+With the linen API, users can define arbitrary variable collections and PRNG sequences.
+Each variable in a collection is lifted in the same way.
+
+Collections are typically given a semantically meaningful name like "params" or "batch_stats" rather than a general purpose name like "state".
+Because collections carry semantic meaning we can decide at the transformation level how each collection should be lifted.
+For example, we might decide to share parameters if we want to vectorize add a batch dimension to a model. 
+
+At the same time we can write generic code that uses transformations without knowing exactly what kind of variables the submodules will create.
+Collections thus strike a balance between fine-grained control and generality.
+We also avoid brittle string matching code that loops over all variables and tries to split up collections in an ad-hoc way based on
+naming conventions like: does this variable name have a prefix "kernel". 
+If more fine-grained control is necessary a user can simply split up a set of variables over multiple collections that should be handled differently.
+
+
+### Transformation vs variable control
+
+
+Lifting behavior could be defined either at the transformation level or during variable definition.
+We use transformation level definitions of lifting behavior.
+The reason for this choice is that there are many different transformations with various behaviors.
+For example: `vmap` has broadcasted and vectorized arguments, while `scan` has scan, carry, and broadcast arguments.
+A variable would have to define its behavior for all these transformations otherwise a `Module` would not be compatible with
+these tranformations. Or alternatively, we would have to make default decisions for how transformations are handled.
+However, this could lead to silent bugs becuase the behavior might not actually be valid given the users intent. 
+
+The lift package also provides a general purpose `transform` which allows an arbitrary function to transforms a variable collection.
```

Comment:
Suggestion:

- "...function to transform a variable collection" (instead of "transforms")

```suggestion
The lift package also provides a general purpose `transform`, which allows an arbitrary function to transform a variable collection.
```